capa/features/extractors/ghidra/insn.py

```diff
@@ -0,0 +1,345 @@
+        push offset aAcr     ; "ACR  > "
+    """
+    data_ref = OperandType.ADDRESS | OperandType.SCALAR  # DAT_* or s_*
+    ref = insn.getAddress()
+    for i in range(insn.getNumOperands()):
+        if insn.getOperandType(i) == data_ref:
+            ref = insn.getAddress(i)
+
+    if ref != insn.getAddress():
+        ghidra_dat = getDataAt(ref)  # type: ignore [name-defined] # noqa: F821
+        if ghidra_dat and ghidra_dat.hasStringValue():
+            yield String(ghidra_dat.getValue()), AbsoluteVirtualAddress(insn.getAddress().getOffset())
+
+
+def extract_insn_mnemonic_features(insn) -> Iterator[Tuple[Feature, Address]]:
+    """parse instruction mnemonic features"""
+    yield Mnemonic(insn.getMnemonicString().lower()), AbsoluteVirtualAddress(insn.getAddress().getOffset())
+
+
+def extract_insn_obfs_call_plus_5_characteristic_features(insn) -> Iterator[Tuple[Feature, Address]]:
+    """
+    parse call $+5 instruction from the given instruction.
+    """
+
+    if not insn.getMnemonicString().startswith("CALL"):
+        return
+
+    code_ref = OperandType.ADDRESS | OperandType.CODE
+    ref = insn.getAddress()
+    for i in range(insn.getNumOperands()):
+        if insn.getOperandType(i) == code_ref:
+            ref = insn.getAddress(i)
+
+    if insn.getAddress().add(5) == ref:
+        yield Characteristic("call $+5"), AbsoluteVirtualAddress(insn.getAddress().getOffset())
+
+
+def extract_insn_segment_access_features(insn) -> Iterator[Tuple[Feature, Address]]:
+    """parse instruction fs or gs access"""
+    insn_str = insn.toString()
+
+    if "FS:" in insn_str:
+        yield Characteristic("fs access"), AbsoluteVirtualAddress(insn.getAddress().getOffset())
+
+    if "GS:" in insn_str:
+        yield Characteristic("gs access"), AbsoluteVirtualAddress(insn.getAddress().getOffset())
+
+
+def extract_insn_peb_access_characteristic_features(insn) -> Iterator[Tuple[Feature, Address]]:
+    """parse instruction peb access
+
+    fs:[0x30] on x86, gs:[0x60] on x64
+
+    """
+    insn_str = insn.toString()
+    if insn_str.startswith("PUSH") or insn_str.startswith("MOV"):
+        if "FS:[0x30]" in insn_str or "GS:[0x60]" in insn_str:
+            yield Characteristic("peb access"), AbsoluteVirtualAddress(insn.getAddress().getOffset())
+
+
+def extract_insn_cross_section_cflow(insn) -> Iterator[Tuple[Feature, Address]]:
+    """inspect the instruction for a CALL or JMP that crosses section boundaries"""
+
+    code_ref = OperandType.ADDRESS | OperandType.CODE
+    data_ref = OperandType.ADDRESS | OperandType.DATA
+
+    insn_str = insn.getMnemonicString()
+    if not (insn_str.startswith("CALL") or insn_str.startswith("J")):
+        return
+
+    this_mem_block = getMemoryBlock(insn.getAddress()).getName()  # type: ignore [name-defined] # noqa: F821
+
+    # assume only CALLs or JMPs are passed
+    ref_type = insn.getOperandType(0)
+    if ref_type != code_ref:
+        if ref_type != data_ref:
+            return
```

Comment:
I found that on instructions like `JMP RCX` I can't pull an address from the register, and the call `insn.getAddress(0)` will end up returning a NoneType. So, I think the explicit checking here is needed to ensure that we can actually pull the address.

Reply:
I've found that when the operand is a register i.e. `JMP RAX` I can't actually pull the address from there. There is an api call such as insn.getRegister(0).getAddress(), but it returns a fake address that represents the register, itself instead of trying to resolve an actual address that may be stored in there. 

However, I can adjust this check to ignore CALLs and JMPs that contain registers as operands instead. I think that would be more simple and slightly faster.

Reply:
This issue actually sent me down a big research rabbit hole into Ghidra and how it dereferences addresses. I found that this snippet:

```python3
for ref in insn.getReferencesFrom():
     reference = ref.getToAddress()
```

Seems like it would handle all cases. But, it would miss on a lot of external functions, especially external library or statically-linked ones, so in a few functions, I've implemented some custom dereference handling in order to avoid things like returning fake addresses (like 0x10) as well as catching those that are missed by the api.